libraries/AP_HAL/UARTDriver.cpp

```diff
@@ -0,0 +1,49 @@
+#include "UARTDriver.h"
+
+#if HAL_LOG_UART_STATS
+
+#include <AP_Logger/AP_Logger.h>
+
+extern const AP_HAL::HAL &hal;
+
+void AP_HAL::UARTDriver::update_logging_all()
+{
+    hal.uartA->update_logging();
+    hal.uartB->update_logging();
+    hal.uartC->update_logging();
+    hal.uartD->update_logging();
+    hal.uartE->update_logging();
+    hal.uartF->update_logging();
+    hal.uartG->update_logging();
+    hal.uartH->update_logging();
+}
+
+void AP_HAL::UARTDriver::update_logging()
+{
+    struct Stats stats_copy;
+    {
+        WITH_SEMAPHORE(_stats.sem);
+        stats_copy = _stats;
+
+        // reset statistics
+        memset(&_stats, 0, sizeof(_stats));
+        _stats.rx_buffer_min = -1; // using integer wrap
+        _stats.tx_buffer_min = -1; // using integer wrap
+    }
+
+    const struct log_UART pkt = {
+    LOG_PACKET_HEADER_INIT(LOG_UART_MSG),
+    time_us                : AP_HAL::micros64(),
+    instance               : (uint8_t)_instance,
+    rxq_min                : stats_copy.rx_buffer_min,
+    rxq_max                : stats_copy.rx_buffer_max,
+    rxq_avg                : (stats_copy.rx_buffer_bytes_available_count == 0) ? (uint16_t)0 : uint16_t(stats_copy.rx_buffer_bytes_available_sum/stats_copy.rx_buffer_bytes_available_count),
```

Comment:
It's a style thing, but casting the whole return from the touple to uint16_t is a bit less verbose